File: 01/main.py
import sys
import numpy as np


def brute_force(start, rots, count_all=True):
    count = 0
    pos = start
    for rot in rots:
        for step in range(1, abs(rot)+1):
            pos = (pos + np.sign(rot)) % 100
            if count_all:
                if pos == 0:
                    count += 1
        if not count_all:
            if pos == 0:
                count += 1
    return count


def main():
    start = 50

    input_file = sys.argv[1]
    with open(input_file) as f:
        rots = f.readlines()

    def parse_rot(s):
        sign = s[0]
        num = int(s[1:])
        if sign == "L":
            return -num
        return num

    rots = np.array([parse_rot(rot) for rot in rots])

    # part one
    # count number of times dial ends at zero
    raw_positions = np.cumsum(np.hstack([[start,], rots]))
    dial_positions = raw_positions % 100
    end_at_zero = dial_positions == 0
    print(np.sum(end_at_zero))
    print(brute_force(start, rots, count_all=False))
    print()

    # part two
    # count number of times dial passes zero
    print(brute_force(start, rots))


if __name__ == "__main__":
    main()

File: 01/main.go
package main

import (
	"bufio"
	"fmt"
	"os"
)

func main() {
	var filename string
	if len(os.Args) < 2 {
		filename = "input"
	} else {
		filename = os.Args[1]
	}

	// open file
	file, err := os.Open(filename)
	if err != nil {
		fmt.Println(err)
		return
	}
	defer file.Close()

	// read and parse input
	var rots []int
	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		line := scanner.Text()
		var rot int
		fmt.Sscanf(line[1:], "%d", &rot)
		if line[:1] == "L" {
			rot = -rot
		}
		rots = append(rots, rot)
	}
	if err := scanner.Err(); err != nil {
		fmt.Println(err)
	}

	var pos = 50 // starting position
	var count_all, count_end int
	for _, rot := range rots {
		step := 1
		if rot < 0 {
			step = -step
			rot = -rot
		}
		for i := 0; i < rot; i++ {
			pos = (pos + step) % 100
			if pos == 0 {
				count_all++
			}
		}
		if pos == 0 {
			count_end++
		}
	}
	// part 1
	fmt.Println(count_end)
	// part 2
	fmt.Println(count_all)
}
